Sources/LeafKit/LeafRenderer.swift

```diff
@@ -60,40 +26,82 @@ public final class LeafRenderer {
     let file: NonBlockingFileIO
     let eventLoop: EventLoop
     
+    // TODO: More Cache Options
+    let cache: LeafCache = Cache()
+    
     public init(
         config: LeafConfig,
         threadPool: BlockingIOThreadPool,
         eventLoop: EventLoop
-        ) {
+    ) {
         self.config = config
         self.file = .init(threadPool: threadPool)
         self.eventLoop = eventLoop
     }
     
     public func render(path: String, context: [String: LeafData]) -> EventLoopFuture<ByteBuffer> {
         let path = path.hasSuffix(".leaf") ? path : path + ".leaf"
-        return self.file.openFile(path: config.rootDirectory + path, eventLoop: self.eventLoop).flatMap { res in
-            return self.file.read(
-                fileRegion: res.1, allocator: ByteBufferAllocator(),
-                eventLoop: self.eventLoop
-                ).flatMapThrowing { buffer in
-                    try res.0.close()
-                    return buffer
-            }
-            }.flatMapThrowing { template in
-                return try self.render(template: template, context: context)
+        let expanded = config.rootDirectory + path
+        let document = fetch(path: expanded)
+        return document.flatMapThrowing { document in
+            throw "todo: serialize document w/ context"
         }
     }
     
-    public func render(template: ByteBuffer, context: [String: LeafData]) throws -> ByteBuffer {
-        var lexer = LeafLexer(template: template)
-        let tokens = try lexer.lex()
-        var parser = LeafParser(tokens: tokens)
-        let raw = try parser.parse()
+    private func fetch(path: String) -> EventLoopFuture<ResolvedDocument> {
+        let path = path.hasSuffix(".leaf") ? path : path + ".leaf"
+        let expanded = config.rootDirectory + path
+        return cache.load(path: expanded, on: eventLoop).flatMap { cached in
+            guard let cached = cached else { return self.read(file: path) }
+            return self.eventLoop.makeSucceededFuture(cached)
+        }
+    }
+    
+    private func read(file: String) -> EventLoopFuture<ResolvedDocument> {
+        let raw = readBytes(file: file)
         
+        let syntax = raw.flatMapThrowing { raw -> [Syntax] in
+            var lexer = LeafLexer(template: raw)
+            let tokens = try lexer.lex()
+            var parser = LeafParser(tokens: tokens)
+            return try parser.parse()
+        }
         
-        #warning("TODO: resolve import / extend / static embed")
-        throw "todo: serialize"
+        return syntax.flatMap { syntax in
+            let dependencies = self.readDependencies(syntax.dependencies)
```

Comment:
Maybe `syntax.dependencies` could be declared here or as a part of `ExtendResolver` somehow? That functionality seems like it will only be used by this code, so having to hunt down the extension is not ideal.